README.processor_trace

```diff
@@ -0,0 +1,100 @@
+Processor trace background information
+--------------------------------------
+
+Processor trace allows instruction and/or data information to be
+traced. Further details of how this is achieved can be found in the
+"Efficient Trace for RISC-V" document which is available from here:
+
+https://github.com/riscv-non-isa/riscv-trace-spec/blob/main/riscv-trace-spec.pdf
+
+In addition python refence models of the trace encoder and trace
+decoder can be found here:
+
+https://github.com/riscv-non-isa/riscv-trace-spec/tree/main/referenceFlow
+
+Processor trace CSV formats
+---------------------------
+
+Both instruction and data trace use a CSV format for the spike
+generated output.
+
+Instruction trace
+-----------------
+
+The header line generated by spike is:
+"VALID,ADDRESS,INSN,PRIVILEGE,EXCEPTION,ECAUSE,TVAL,INTERRUPT"
+
+The first column is a constant value 1 (one) present in all rows of
+the output.
+
+The remaining columns are named by their meaning. Identical naming of
+columns in spike files is not required, whereas the correct number of
+columns and the correct data types are required.
+
+VALID: Always set to 1
+
+ADDRESS: Address of the instruction (%lx format, reg_t)
+
+INSN: Instruction (%lx format, uint64_t)
+
+PRIVILEGE: Privilege level (%x format, uint8_t)
+
+EXCEPTION: Whether an exception occurred (%x format, uint8_t)
+
+ECAUSE: The exception value when an exception occurred (%lx format, reg_t)
+
+TVAL: The trap value (%lx format, reg_t)
+
+INTERRUPT: Whether an interrupt occurred (%x format, uint8_t)
+
+Data trace
+----------
+
+The header line generated by spike is:
+"DRETIRE,DTYPE,DADDR,DSIZE,DATA"
+
+DRETIRE: Whether the instruction retired. Currently this is always set
+to 1.
+
+DTYPE: This is the data access type that has been generated. (See
+Table 4.12 in the trace specification) (%d format, dtype_t)
+  0    Load
+  1    Store
+  2    reserved
+  3    reserved
+  4    CSR read-write
+  5    CSR read-set
+  6    CSR read-clear
+  7    reserved
+  8    Atomic swap
+  9    Atomic add
+  10   Atomic AND
+  11   Atomic OR
+  12   Atomic XOR
+  13   Atomic max
+  14   Atomic min
+  15   Conditional store failure
+
+DADDR: Address of the data access. (%lx format, reg_t)
+
+DSIZE: The data transfer is 2^DSIZE bytes. (%ld format, size_t)
+
+DATA: The data value or values transferred. (hex format, length is
+dependent on DSIZE)
+
+Testing
+-------
+
+A small collection of tests are included which run spike to generate instruction and data trace
+file and then checks the output against expected files. Both 64-bit and 32-bit binaries are
```

Comment:
Where did the expected files originate? Has there been any independent validation of their contents?

Reply:
They were produced by a run through a version of spike. The validation that has been done is twofold:
- the traces have been run through various models of the trace encoder and decoder within Siemens and by their customers, and it has been checked that the original trace was reconstructable.
- for some of the smaller files the data has been scrutinized within Siemens and by their customers to check that it agrees with hardware models. In particular, Seagate has contributed a "torture" test example as part of their RISCV development work (https://github.com/riscv-non-isa/riscv-trace-spec/blob/main/referenceFlow/tests/examples/br_j_asm_test/test.S).

Reply:
My concern is how we keep these traces up to date over time. As Spike changes, these traces might need to change too.

I guess it wouldn't be too hard to eyeball any differences and then manually copy the new traces over the golden traces (assuming the changes are intentional)? Could/should we document that process?

Should we run these tests as part of the GitHub CI? Otherwise people will submit PRs and not even realize they've broken the tracing.

Reply:
I share your concern over using expected files for CI.
It's hard to quantify, but our experience with the instruction trace files, over a number of years (data trace was much more recent), is that the trace output has not changed as we have updated the patch file to keep it up to date with your upstream changes. Given the fixed elf file input, I would expect that there should be no reason for the trace to change.

Looking at  some of the core instruction trace code
```
      proc_trace->set_addr(state.pc);
      proc_trace->set_insn(insn, bits);
      proc_trace->set_priv(state.prv);
```
e.g. if the encoding or data type of state.prv were to change, then I would expect that there would be large numbers of differences against the expected files (or it might not even compile?).

Regarding documenting how to deal with CI failures:
- the traces include the start-up code that is part of spike; if that changes then in terms of content or location, then the expected files will need to be updated
- if the data types of e.g. internal state of processors was to be changed, then the C++ formatted output of the trace data may well need to be changed
- given that the elf file input to the tests is fixed (and is not subject to changes in the toolchain), I would expect that the instructions addresses will not change (since proxy kernel operation is not supported for trace)

Reply:
I suppose in that case there would probably need to be a re-compilation of the source and the new output from running the instruction and data trace taken as the expected output. Are you saying that, in the case that you have cited, the old elf files would not run using an updated spike?

Reply:
In the case I'm imagining the old elf files would not run the same way anymore, and would not give good trace coverage anymore. Maybe this is far-fetched.

There is also the case where we discover a problem with the tracing, and we want to expand coverage to make this problem visible, which means modifying one of the elfs, or adding a new elf.

I want to avoid the provenance of these elf files being a mystery to anyone new coming into this project. "I made these by hand once and I'm assuming they will never need to be changed" is not generally a good plan.

Reply:
I agree that lack of provenance is not a good plan to build in. I'll create something to provide some provenance.

Reply:
I've added extra information about how to rebuild the various test binaries. In particular, I've added the script and C entry point code required to build the embench code for use with spike.

Reply:
I was expecting, that SPIKE output itself can be used to check if encoder+decoder will produce desired PC sequence.
Spike patch is producing PC sequence (+ other details). If we extract 'PC-only flow', we can check if encoder+decoder are is producing same PC flow. Using such a technique ANY file/program which can be run by spike can be used to test trace. Similar technique can be used to validate data trace - trace should reconstruct all memory transactions (same addresses and data).